api/envoy/config/rbac/v2alpha/rbac.proto

```diff
@@ -60,96 +67,76 @@ message RBAC {
 // Policy specifies a role and the principals that are assigned/denied the role.
 message Policy {
  // Required. The set of permissions that define a role.
- repeated Permission permissions = 1 [(validate.rules).repeated .min_items = 1];
+ repeated Permission permissions = 1 [(validate.rules).repeated.min_items = 1];
 
  // Required. List of principals that are assigned/denied the role based on “action”.
- repeated Principal principals = 2 [(validate.rules).repeated .min_items = 1];
+ repeated Principal principals = 2 [(validate.rules).repeated.min_items = 1];
 }
 
-// Specifies how to match an entry in a map.
-message MapEntryMatch {
-  // The key to select an entry from the map.
-  string key = 1;
+// Permission defines a permission to access the service.
+message Permission {
 
-  // A list of matched values.
-  repeated envoy.type.StringMatch values = 2;
-}
+  message Set {
+    repeated Permission rules = 1 [(validate.rules).repeated.min_items = 1];
+  }
 
-// Specifies how to match IP addresses.
-message IpMatch {
-  // IP addresses in CIDR notation.
-  repeated envoy.api.v2.core.CidrRange cidrs = 1;
-}
+  oneof rule {
+    option (validate.required) = true;
 
-// Specifies how to match ports.
-message PortMatch {
-  // Port numbers.
-  repeated uint32 ports = 1;
-}
+    // A set of rules that all must match in order to define the action.
+    Set and_rules = 1;
 
-// Permission defines a permission to access the service.
-message Permission {
-  // Optional. A list of HTTP paths or gRPC methods.
-  // gRPC methods must be presented as fully-qualified name in the form of
-  // packageName.serviceName/methodName.
-  // If this field is unset, it applies to any path.
-  repeated envoy.type.StringMatch paths = 1;
-
-  // Required. A list of HTTP methods (e.g., "GET", "POST").
-  // If this field is unset, it applies to any method.
-  repeated string methods = 2;
-
-  // Definition of a custom condition.
-  message Condition {
-    oneof condition_spec {
-      // Header match. This matches to the "request.http.headers" field in
-      // ":ref: `AttributeContext <envoy_api_msg_service.auth.v2alpha.AttributeContext>`.
-      // The map key is the header name. The header specifies how the service is accessed.
-      MapEntryMatch header = 1;
-
-      // Destination IP addresses.
-      IpMatch destination_ips = 2;
-
-      // Destination ports.
-      PortMatch destination_ports = 3;
-    }
-  }
+    // A set of rules where at least one must match in order to define the action.
+    Set or_rules = 2;
 
-  // Optional. Custom conditions.
-  repeated Condition conditions = 3;
+    // When any is set, it matches any action.
+    bool any = 3 [(validate.rules).bool.const = true];
```

Comment:
If "any" is unset or set to false, and there is no permission defined, what is the behavior?

Reply:
The oneof has a required validation rule. You would not be able to emit a permission/principal without a rule set. This was intended to make the intent explicit (instead of by convention).

Reply:
Sounds good. Thanks.